text/0517-io-os-reform.md

```diff
@@ -473,12 +658,246 @@ throughout IO, we can go on to explore the modules in detail.
 ### `core::io`
 [core::io]: #coreio
 
-> To be added in a follow-up PR.
+The `io` module is split into the parts that can live in `libcore`
+(most of it) and the parts that are added in the `std::io`
+facade. Being able to move components into `libcore` at all is made
+possible through the use of
+[associated error types](#revising-reader-and-writer) for `Reader` and
+`Writer`.
+
+#### Adapters
+[Adapters]: #adapters
+
+The current `std::io::util` module offers a number of `Reader` and
+`Writer` "adapters". This RFC refactors the design to more closely
+follow `std::iter`. Along the way, it generalizes the `by_ref` adapter:
+
+```rust
+trait ReaderExt: Reader {
+    // already introduced above
+    fn bytes(&mut self) -> Bytes<Self> { ... }
+    fn chars<'r>(&'r mut self) -> Chars<'r, Self, Err> { ... }
+
+    // Reify a borrowed reader as owned
+    fn by_ref<'a>(&'a mut self) -> ByRef<'a, Self> { ... }
+
+    // Read everything from `self`, then read from `next`
+    fn chain<R: Reader>(self, next: R) -> Chain<Self, R> { ... }
+
+    // Adapt `self` to yield only the first `limit` bytes
+    fn take(self, limit: u64) -> Take<Self> { ... }
+
+    // Whenever reading from `self`, push the bytes read to `out`
+    fn tee<W: Writer>(self, out: W) -> Tee<Self, W> { ... }
```

Comment:
@aturon and I had the idea of something like this which may solve this problem:

```rust
enum TeeError<R, W> {
    ReadError(R),
    WriteError(usize, W),
}

impl<E> FromError<TeeError<E, E>> for E {
    fn from_error(tee_error: TeeError<E, E>) -> E { ... }
}
```

@aturon, perhaps that could be added to this RFC (albeit it's getting quite large!)